app/services/action_handler.rb

```diff
@@ -0,0 +1,17 @@
+class ActionHandler < BaseService
+  def initialize(payload:, event:, event_type:)
+    @payload = payload
+    @event = event
+    @event_type = event_type
+  end
+
+  def call
+    resolve
+  end
+
+  private
+
```

Comment:
I don't get why would we need to implement this.
```
def handleable?
  raise "Subclass #{self.class.name} must implement the method :handleable?"
end
```
Rails by himself has his own Exception when a method is not implemented.

And this:
```
def handled_actions
  self.class.ACTIONS
end

def handleable?
  handled_actions.include?(@payload['action'])
end
```
Won't work, because we can't access to constants with `self.class.ACTIONS`.

Reply:
Having a different exception from `missing_method` is solely for documentation purposes.

A missing method can be raised for different reasons:

- the sender could be sending the wrong message to the right object
- the sender could be sending the right message to the wrong object
- the sender could be sending the right message to the right object but the error is on the receiver side

A `raise ObjectShouldRespondToMessage.new` documents that a missing method would occur in the context of third scenario. If a developer is creating a new class that should comply with the ActionHandler protocol she would get the error stating that the method should be implemented and will have a bit more context to implement the class.

From the `ActionHandler` protocol side it allows a reader to know which methods does the ActionHandler expects to be implemented in a different place.

Probably a better implementation would be

```
def handleable?
  raise ObjectShouldRespondToMessage.new
end
```

and yet a more sophisticated approach could be something like

```
class ActionHandler < BaseService
  extend Interface

  object_should_respond_to :handleable?
end
```

The reasons are arguable, of course.

Reply:
I missed that ACTIONS would not be accessible, sorry.

Still may be it would be possible to create an accessor method in the subclass:

```
class PullRequest
  # Returns an Array with all the payload actions the receiver handles.
   def handled_actions
      ACTIONS
   end
end
```

and the super class would use it like

```
class ActionHandler < BaseService
  ...
  private

  # Returns an Array with all the payload actions the receiver handles.
  # Each action is expected to be a String.
  def handled_actions
    raise ObjectShouldRespondToMessageError.new
  end

  # Returns true if the receiver handles the action of the payload it holds, false otherwise.
  def handleable?
    handled_actions.include?(@payload['action'])
  end

  # Handles the action of the payload it holds.
  # Assumes the action is :handleable?. This assumption must be validated prior to calling
  # this method.
  def handle_action
    send(@payload['action'])
  end
end
```

Reply:
I am trying to understand what you said, what do you want to implement is an `Abstract Factory`?
Because otherwise we are overloading the app with unnecessary code, Ruby code should be self-document. Also, we are skipping Rails exception control.
If you are trying to implement Abstract Factory, Rails already does it in the `Base` class. And if you want to log it, you should use a `Rescue` clause with a custom message.

Reply:
Nop, no factory.

`PullRequest`, `ReviewComment` and `Review` classes all implement the methods

```
def resolve
  return unless handleable?

  handle_action
end

def handleable?
  ACTIONS.include?(@payload['action'])
end
```

All of these classes also inherit from `ActionHandler`.

I was suggesting to remove these methods from the subclasses and move them to `ActionHandler`:

```
class ActionHandler < BaseService
  # ...

  def resolve
    return unless handleable?
  
    handle_action
  end

  def handleable?
    ACTIONS.include?(@payload['action'])
  end
end
```
replacing the reference to the ACTIONS constant by a method named :handled_actions

```
class ActionHandler < BaseService
  # ...

  def resolve
    return unless handleable?
  
    handle_action
  end

  def handleable?
    handled_actions.include?(@payload['action'])
  end
end
```

```
class PullRequest < ActionHandler
  # ..

  def handled_actions
    ACTIONS
  end
end

# same for `ReviewComment` and `Review`
```